File: lib/math_utils.py
import math
import random
import sys

def addNormalizedValues(arr, key, nkey):
    if len(arr) < 1:
        return arr
    values = [v[key] for v in arr]
    range = (min(values), max(values))
    for i, entry in enumerate(arr):
        arr[i][nkey] = norm(entry[key], range)
    return arr

def ceilInt(value):
    return int(math.ceil(value))

def isInteger(value):
    return isinstance(value, int)

def isNumeric(value):
    if value == "":
        return False
    response = True
    try:
        num = float(value)
    except ValueError:
        response = False
    return response

def lerp(ab, amount):
    a, b = ab
    return (b-a) * amount + a

def lim(value, ab=(0, 1)):
    a, b = ab
    return max(a, min(b, value))

def norm(value, ab, limit=False):
    a, b = ab
    n = 0.0
    if (b - a) != 0:
        n = 1.0 * (value - a) / (b - a)
    if limit:
        n = lim(n)
    return n

def parseNumber(string, alwaysFloat=False, invalidValue="self"):
    try:
        num = float(string)
        if "." not in str(string) and not alwaysFloat:
            num = int(string)
        return num
    except ValueError:
        if invalidValue == "self":
            return string
        else:
            return invalidValue

def parseNumbers(arr):
    for i, item in enumerate(arr):
        if isinstance(item, (list,)):
            for j, v in enumerate(item):
                arr[i][j] = parseNumber(v)
        else:
            for key in item:
                if key != "id":
                    arr[i][key] = parseNumber(item[key])
    return arr

def printProgress(step, total, prepend=""):
    sys.stdout.write('\r')
    sys.stdout.write("%s%s%%" % (prepend, round(1.0*step/total*100,2)))
    sys.stdout.flush()

def randomPointInSphere(center, radius, seed=None):
    if seed is not None:
        random.seed(seed)
    phi = random.uniform(0, 2*math.pi)
    if seed is not None:
        random.seed(seed+1)
    costheta = random.uniform(-1, 1)
    if seed is not None:
        random.seed(seed+2)
    u = random.random()
    theta = math.acos(costheta)
    r = radius * (u ** (1. / 3))

    x = r * math.sin(theta) * math.cos(phi)
    y = r * math.sin(theta) * math.sin(phi)
    z = r * math.cos(theta)

    return (center[0]+x, center[1]+y, center[2]+z)

def randomUniform(fromValue=0, toValue=1.0, seed=None):
    if seed is not None:
        random.seed(seed)
    return random.uniform(fromValue, toValue)

def roundInt(value):
    return int(round(value))

# East = 0 degrees
def translatePoint(x, y, distance, angle, radians=False):
    rad = angle if radians else math.radians(angle)
    x2 = x + distance * math.cos(rad)
    y2 = y + distance * math.sin(rad)
    return (x2, y2)

File: js/utils/math_utils.js
(function() {
  window.MathUtil = {};

  MathUtil.clamp = function(value, min, max) {
    value = Math.min(value, max);
    value = Math.max(value, min);
    return value;
  };

  MathUtil.degToRad = function (degrees) {
    return degrees * Math.PI / 180;
  };

  MathUtil.ease = function(n){
    return (Math.sin((n+1.5)*Math.PI)+1.0) / 2.0;
  };

  MathUtil.lerp = function(a, b, percent) {
    return (1.0*b - a) * percent + a;
  };

  MathUtil.norm = function(value, a, b){
    var denom = (b - a);
    if (denom > 0 || denom < 0) {
      return (1.0 * value - a) / denom;
    } else {
      return 0;
    }
  };

  MathUtil.randomBetween = function(a, b){
    var delta = b - a;
    if (delta <= 0) return 0;
    return Math.random() * delta + a;
  };

  MathUtil.randomPointInCylinder = function(baseCenter, radius, height) {
    var s = Math.random();
    var theta = MathUtil.randomBetween(0, 2*Math.PI);
    var r = Math.sqrt(s) * radius;
    var x = r * Math.cos(theta);
    var y = r * Math.sin(theta);
    var z = MathUtil.randomBetween(0, height);
    return [baseCenter[0]+x, baseCenter[1]+y, baseCenter[2]+z];
  };

  MathUtil.randomPointInSphere = function(center, radius) {
    var phi = MathUtil.randomBetween(0, 2*Math.PI);
    var costheta = MathUtil.randomBetween(-1, 1);
    var u = Math.random();
    var theta = Math.acos(costheta);
    var r = radius * Math.pow(u, (1./3));
    var x = r * Math.sin(theta) * Math.cos(phi)
    var y = r * Math.sin(theta) * Math.sin(phi)
    var z = r * Math.cos(theta)
    return [center[0]+x, center[1]+y, center[2]+z];
  };


})();
